delta-app/src/main/java/io/cdap/delta/store/AssessmentServiceClient.java

```diff
@@ -38,71 +38,65 @@
 import java.time.temporal.ChronoUnit;
 import javax.annotation.Nullable;
 import javax.ws.rs.HttpMethod;
+import javax.ws.rs.core.MediaType;
 
 /**
  * Class to handle Api calls to System Service
  */
 public class AssessmentServiceClient {
 
-  ServiceDiscoverer serviceDiscoverer;
+  private final ServiceDiscoverer serviceDiscoverer;
+  private RetryPolicy retryPolicy;
+  private static final Gson GSON = new Gson();
 
   public AssessmentServiceClient(ServiceDiscoverer context) {
     this.serviceDiscoverer = context;
+    setUpRetryPolicy();
   }
 
-  public String retryableApiCall(String connectionUrl, String method, @Nullable String requestBody) {
+  public <T> T retryableApiCall(String connectionUrl, String method, @Nullable String requestBody,
+                                @Nullable Class<T> type) {
 
-    RetryPolicy retryPolicy = new RetryPolicy()
-      .withMaxAttempts(5)
-      .withMaxDuration(Duration.of(10, ChronoUnit.SECONDS))
-      .withDelay(Duration.of(200, ChronoUnit.MILLIS))
-      .withJitter(0.20D);
-
-    FailsafeExecutor<String> failsafe = Failsafe.with(retryPolicy);
+    FailsafeExecutor<T> failsafe = Failsafe.with(retryPolicy);
 
     return failsafe.get(() -> {
       if (method.equals(HttpMethod.GET)) {
-        return executeGetCall(getHttpUrlConnection(connectionUrl));
+        return executeGetCall(getHttpUrlConnection(connectionUrl), type);
       } else if (method.equals(HttpMethod.POST) || method.equals(HttpMethod.PUT)) {
-        return executeCallWithBody(getHttpUrlConnection(connectionUrl), method, requestBody);
+        return executeCallWithBody(getHttpUrlConnection(connectionUrl), method, requestBody, type);
       }
       throw new IllegalStateException("The given Http Method Doesn't match with available methods");
     });
-
   }
 
-  private String executeGetCall(HttpURLConnection urlConn) throws IOException {
+  private <T> T executeGetCall(HttpURLConnection urlConn, Class<T> type) throws IOException {
     urlConn.setRequestMethod(HttpMethod.GET);
-    return validateAndRetrieveContent(urlConn);
+    return validateAndRetrieveContent(urlConn, type);
   }
 
-  private String executeCallWithBody(HttpURLConnection urlConn, String method, String body)
+  private <T> T  executeCallWithBody(HttpURLConnection urlConn, String method, String body, Class<T> type)
     throws IOException {
 
-    OutputStream os = null;
     OutputStreamWriter osw = null;
+    urlConn.setDoOutput(true);
+    urlConn.setRequestMethod(method);
 
     try {
-      urlConn.setDoOutput(true);
-      urlConn.setRequestMethod(method);
-      os = urlConn.getOutputStream();
-      osw = new OutputStreamWriter(os, "UTF-8");
+      osw = new OutputStreamWriter(urlConn.getOutputStream(), "UTF-8");
```

Comment:
does

```
try (OutputStreamWriter osw = ...) {

}
```

not work? I thought it was an auto-closeable?

Reply:
Yes it works. I missed it the first time. 
Rectified it